File: R/uniform.R
#' Sample from uniform distribution on the interval from `min` to `max`.
#'
#' @param n Number of samples to draw.
#' @param min Minimum value of the uniform distribution.
#' @param max Maximum value of the uniform distribution.
#' @param random_seed Random seed for reproducibility.
#' @param method Underlying C++ sampling method to use ("std", "boost", or "custom").
#'
#' @return A vector of sampled values of length `n`.
#' @export
#'
#' @examples
#' draws <- sample_uniform(n = 10, method = "std")
sample_uniform <- function(n, min = 0, max = 1, random_seed = -1, method = "custom") {
  if (method == "std") {
    return(sample_uniform_std_cpp(n, min, max, random_seed))
  } else if (method == "boost") {
    return(sample_uniform_boost_cpp(n, min, max, random_seed))
  } else if (method == "custom") {
    return(sample_uniform_custom_cpp(n, min, max, random_seed))
  } else {
    stop("Unknown method")
  }
}

File: src/uniform.cpp
#include <cpp11.hpp>
#include <random>
#include <boost/random/mersenne_twister.hpp>
#include <boost/random/uniform_real_distribution.hpp>
#include "utils.h"

[[cpp11::register]]
cpp11::writable::doubles sample_uniform_std_cpp(int n, double min, double max, int random_seed = -1) {
    // Initialize output vector
    cpp11::writable::doubles output(n);

    // Initialize rng
    std::mt19937 gen = create_std_mt19937(random_seed);

    // Initialize and sample from distribution
    std::uniform_real_distribution<double> unif_dist(min, max);
    for (int i = 0; i < n; i++) {
      output[i] = unif_dist(gen);
    }

    return output;
}

[[cpp11::register]]
cpp11::writable::doubles sample_uniform_boost_cpp(int n, double min, double max, int random_seed = -1) {
    // Initialize output vector
    cpp11::writable::doubles output(n);

    // Initialize rng
    boost::random::mt19937 gen = create_boost_mt19937(random_seed);

    // Initialize and sample from distribution
    boost::random::uniform_real_distribution<double> unif_dist(min, max);
    for (int i = 0; i < n; i++) {
      output[i] = unif_dist(gen);
    }

    return output;
}

[[cpp11::register]]
cpp11::writable::doubles sample_uniform_custom_cpp(int n, double min, double max, int random_seed = -1) {
    // Initialize output vector
    cpp11::writable::doubles output(n);

    // Initialize rng
    std::mt19937 gen = create_std_mt19937(random_seed);

    // Sample from the distribution directly by 
    // generating a standard uniform sample and then
    // adjusting to the target min and max range
    double range = max - min;
    for (int i = 0; i < n; i++) {
      output[i] = standard_uniform_draw(gen) * range + min;
    }

    return output;
}
